### PR TITLE
Support ${} style variables in live logs #607

### LINKED ISSUE #607: Support ${} style variables in live logs
In kotlin you can output variables like:
```kotlin
var s = "hello"
println("$s world") //outputs hello world
```

And you can also output variables like:
```kotlin
var s = "hello"
println("${s} world") //outputs hello world
```

I believe they're called string templates and having the ability to use them allows you to write something like:
```kotlin
var s = "He"
println("${s}llo") //outputs Hello
```

---

Currently, the log status bar only supports the first method:
Peek 2022-01-03 16-04

Trying to use templates results in:
Peek 2022-01-03 16-04-34

### LINKED ISSUE #620: Missing command highlighting on partial text matches
When searching for commands you have to input words fully for them to highlight:
Peek 2022-01-04 19-13

It should highlight even when you input partial words.

### PR DESCRIPTION
**Pull request by javamak**

### CODE REVIEW

**Comment by javamak** [comment]
template

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java [hunk 1]
```diff
@@ -120,11 +130,13 @@ public Icon getIcon() {
         }).collect(Collectors.toList());
         lookup = text -> scopeVars.stream()
                 .filter(v -> {
-                    String var = substringAfterLast(" ", text.toLowerCase());
-                    return var.startsWith("$") && !var.substring(1).equals(v) && v.toLowerCase().contains(var.substring(1));
+                    String var = substringAfterLast(SPACE, text.toLowerCase());
+
+                    return var.startsWith(DOLLOR) && !var.substring(1).equals(v) && v.toLowerCase().contains(var.substring(1)) ||
+                           var.startsWith("${") && !var.substring(2).equals(v) && v.toLowerCase().contains(var.substring(2));
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: DOLLOR should be DOLLAR and you've missed a `$` here

**Comment by BFergerson** [comment]
Log template doesn't work when there is no spacing:
Peek 2022-01-04 18-43

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java (hunk 1 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: Constant renamed. I don't understand the second part. I have not missed $. Please correct me

Comment (BFergerson) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: `startsWith("${")`

**Comment by javamak** [comment]
> Log template doesn't work when there is no spacing: Peek 2022-01-04 18-43

Fixed this problem
template_color

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java (hunk 1 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: right now the code checks startswith of both $ or ${ and it works. Please check

**Comment by BFergerson** [comment]
You still haven't actually implemented the template. Just coloring:
Peek 2022-01-05 16-45

---

What's created:
```
[{
  "id" : "24506938-a098-4cc8-ba06-e48b69aec3c6",
  "location" : {
    "source" : "spp.example.webapp.controller.LiveInstrumentController",
    "line" : 36
  },
  "expiresAt" : 1641420012845,
  "hitLimit" : 100,
  "applied" : true,
  "pending" : false,
  "throttle" : {
    "limit" : 1,
    "step" : "SECOND"
  },
  "type" : "LOG",
  "meta" : {
    "original_source_mark" : "f65193ee-bd2c-4ed8-9bd9-a1bafc8cebb9",
    "created_at" : "1641419112856",
    "created_by" : "system",
    "hit_count" : "84",
    "applied_at" : "1641419112976",
    "first_hit_at" : "1641419113340",
    "last_hit_at" : "1641419279339"
  },
  "logFormat" : "hello${s}hello",
  "logArguments" : [ ]
}]
```

What's expected:
```
  "logFormat" : "hello{}hello",
  "logArguments" : [ "s" ]
```

**Comment by BFergerson** [comment]
The formats acceptable should be:
```kt
var s = "test"

val one = "hello $s" //hello {} (arg: s)
val two = "hello$s" //hello{} (arg: s)
val two = "hello${s}hello" //hello{}hello (arg: s)
```

You should be able to combine them too, like:
```kt
hello$s hello $s hello${s}hello //hello{} hello {} hello{}hello (args: s, s, s)
```

**Comment by javamak** [comment]
> You still haven't actually implemented the template. Just coloring: Peek 2022-01-05 16-45
> 
> What's created:
> 
> ```
> [{
>   "id" : "24506938-a098-4cc8-ba06-e48b69aec3c6",
>   "location" : {
>     "source" : "spp.example.webapp.controller.LiveInstrumentController",
>     "line" : 36
>   },
>   "expiresAt" : 1641420012845,
>   "hitLimit" : 100,
>   "applied" : true,
>   "pending" : false,
>   "throttle" : {
>     "limit" : 1,
>     "step" : "SECOND"
>   },
>   "type" : "LOG",
>   "meta" : {
>     "original_source_mark" : "f65193ee-bd2c-4ed8-9bd9-a1bafc8cebb9",
>     "created_at" : "1641419112856",
>     "created_by" : "system",
>     "hit_count" : "84",
>     "applied_at" : "1641419112976",
>     "first_hit_at" : "1641419113340",
>     "last_hit_at" : "1641419279339"
>   },
>   "logFormat" : "hello${s}hello",
>   "logArguments" : [ ]
> }]
> ```
> 
> What's expected:
> 
> ```
>   "logFormat" : "hello{}hello",
>   "logArguments" : [ "s" ]
> ```

fixed now. 
variable_highlight

**Comment by BFergerson** [comment]
Still a no go:
Peek 2022-01-05 18-13


What's sent:
```
"logFormat": "hello$shello hello $s hello hello{}hello",
"logArguments": [
	"s",
	"{s}"
]
```

What's expected:
```
"logFormat": "hello$shello hello {} hello hello{}hello",
"logArguments": [
	"s",
	"s"
]
```

**Comment by BFergerson** [comment]
Please add unit tests to the PR.

**Comment by javamak** [comment]
variable_fix
Fixed the latest scenario and 1 more scenario for pattern hello$tt $tthello ${tt}$tt.

**Comment by javamak** [comment]
> Please add unit tests to the PR.

checking it

**Comment by javamak** [comment]
> > Please add unit tests to the PR.
> 
> checking it

Testcases added. Please check

File: plugin/src/test/kotlin/spp/jetbrains/VariableParserTest.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.junit.jupiter.api.Test
+import spp.jetbrains.sourcemarker.VariableParser
+import java.util.ArrayList;
+import java.util.regex.Pattern
+
+internal class VariableParserTest {
+
+    var variablePattern =
+        Pattern.compile("(\\\$tt|\\\$deleteDate|\\\$executorService|\\\$FIVE_MINUTES_MILLIS|\\\$mapTodos|\\\$GET|\\\$HEAD|\\\$POST|\\\$PUT|\\\$PATCH|\\\$DELETE|\\\$OPTIONS|\\\$TRACE)(?:\\s|$)")
+    var templateVariablePattern =
```
Comment (BFergerson) [inline plugin/src/test/kotlin/spp/jetbrains/VariableParserTest.kt]: Don't hardcode the variable pattern. This should be done by VariableParser.

Each test should define vars and pattern. Something like:
```kotlin
val contextVars = listOf("deleteDate", "executorService")
val pattern = "hello\$deleteDate \${executorService}"
```

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        }
+
+        if (templateVariablePattern != null) {
+            Matcher m = templateVariablePattern.matcher(logPattern);
+            int matchLength = 0;
+            while (m.find()) {
+                String var = m.group(1);
+                logPattern = logPattern.substring(0, m.start() - matchLength)
+                        + logPattern.substring(m.start() - matchLength).replaceFirst(Pattern.quote(var), "{}");
+                //logPattern = logPattern.replaceFirst(Pattern.quote(var), "{}");
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: No commented code. If it's not necessary, delete it.

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java [hunk 4]
```diff
@@ -62,12 +63,21 @@
 
     private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("hh:mm:ss a")
             .withZone(ZoneId.systemDefault());
+    private static final String DOLLAR = "$";
+    private static final String EMPTY = "";
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: Some of these variables seem like they could go to VariablePattern now. `EMPTY` is being duplicated here.

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java [hunk 5]
```diff
@@ -126,14 +140,19 @@ public Icon getIcon() {
 
         if (!scopeVars.isEmpty()) {
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: This logic looks like it should be in VariablePattern

File: plugin/src/main/kotlin/spp/jetbrains/sourcemarker/status/util/AutocompleteField.kt [hunk 6]
```diff
@@ -88,14 +91,19 @@ class AutocompleteField(
         addKeyListener(this)
 
         val sb = StringBuilder("(")
+        val sbt = StringBuilder("(")
```
Comment (BFergerson) [inline plugin/src/main/kotlin/spp/jetbrains/sourcemarker/status/util/AutocompleteField.kt]: This logic looks like it belongs in VariablePattern too.

Comment (javamak) [reply plugin/src/main/kotlin/spp/jetbrains/sourcemarker/status/util/AutocompleteField.kt]: working on it. but again IDE is failing to connect to Source++. Need help

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java (hunk 3 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: removed

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java (hunk 4 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: resolved

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java (hunk 5 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: resolved

File: plugin/src/main/kotlin/spp/jetbrains/sourcemarker/status/util/AutocompleteField.kt (hunk 6 above)
Comment (javamak) [reply plugin/src/main/kotlin/spp/jetbrains/sourcemarker/status/util/AutocompleteField.kt]: resolved

File: plugin/src/main/java/spp/jetbrains/sourcemarker/PluginUI.java [hunk 7]
```diff
@@ -41,7 +41,7 @@
     public static final Color EXPIRY_FOREGROUND_COLOR = Color.decode("#BBBBBB");
     public static final Color SELECT_COLOR_RED = Color.decode("#e1483b");
     public static final Color COMPLETE_COLOR_PURPLE = Color.decode("#9876AA");
-    public static final Color AUTO_COMPLETE_SELECT_BACKGROUND = Color.decode("#1C1C1C");
+    //public static final Color AUTO_COMPLETE_SELECT_BACKGROUND = Color.decode("#1C1C1C");
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/PluginUI.java]: No commented code. Please check all your files

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import static spp.jetbrains.sourcemarker.Constants.EMPTY;
+import static spp.jetbrains.sourcemarker.Constants.PATTERN_CURLY_BRACES;
+
+public class VariableParser {
+
+    private Pattern variablePattern;
+    private Pattern templateVariablePattern;
+
+    public void createPattern(List<String> scopeVars) {
+
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: Unnecessary spacing. Here and other places

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static spp.jetbrains.sourcemarker.Constants.EMPTY;
+import static spp.jetbrains.sourcemarker.Constants.PATTERN_CURLY_BRACES;
+
+public class VariableParser {
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: This class is not thread-safe. It should be stateless

File: plugin/src/main/java/spp/jetbrains/sourcemarker/Constants.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package spp.jetbrains.sourcemarker;
+
+import java.util.regex.Pattern;
+
+public interface Constants {
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/Constants.java]: This class feels superfluous. Everything shared between classes should have moved to VariableParser. Things like `WAITING_FOR_LIVE_LOG_DATA` and `SPACE` aren't used anywhere but LogStatusBar. They have no point in being in a shared place.

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java [hunk 11]
```diff
@@ -106,11 +124,15 @@ public Icon getIcon() {
         }).collect(Collectors.toList());
         lookup = text -> scopeVars.stream()
                 .filter(v -> {
-                    String var = substringAfterLast(" ", text.toLowerCase());
-                    return var.startsWith("$") && !var.substring(1).equals(v) && v.toLowerCase().contains(var.substring(1));
+                    String var = substringAfterLast(Constants.SPACE, text.toLowerCase());
```
Comment (BFergerson) [inline plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: This logic looks to belong to VariableParser.

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java (hunk 9 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: The class is acting as a container for existing variables and its related methods and the object is not shared. Looks fine to me. Please suggest.

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java (hunk 8 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: resolved

File: plugin/src/main/java/spp/jetbrains/sourcemarker/PluginUI.java (hunk 7 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/PluginUI.java]: resovled

File: plugin/src/main/java/spp/jetbrains/sourcemarker/Constants.java (hunk 10 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/Constants.java]: Removed and constants moved to classes wherever its primarily used

File: plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java (hunk 11 above)
Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/status/LogStatusBar.java]: refactored.

File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java (hunk 9 above)
Comment (BFergerson) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: A bit odd you're not aware of thread-safety. See: https://www.baeldung.com/java-thread-safety

Comment (javamak) [reply plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java]: Resolved. Please check

**Pull request merged**